Instruments/source/CacheTupleProducer.cxx

```diff
@@ -0,0 +1,161 @@
+        for(unsigned n = 0; n < selections.size(); ++n){
+            SignalObjectSelector signalObjectSelector(Parse<analysis::SignalMode>(selections.at(n)));
+            signalObjectSelectors.push_back(signalObjectSelector);
+        }
+
+        std::vector<std::string> sources = SplitValueList(args.unc_source(),false,",");
+        for(unsigned n = 0; n < sources.size(); ++n){
+            UncertaintySource source = Parse<analysis::UncertaintySource>(sources.at(n));
+            unc_sources.push_back(source);
+        }
+    }
+
+    void Run()
+    {
+        std::cout << boost::format("Processing input file '%1%' into output file '%2%' using %3% selection.\n")
+                   % args.input_file() % args.output_file() % args.selection();
+
+        auto originalFile = root_ext::OpenRootFile(args.input_file());
+        auto outputFile = root_ext::CreateRootFile(args.output_file());
+        auto originalTuple = ntuple::CreateEventTuple(args.tree_name(),originalFile.get(),true,ntuple::TreeState::Full);
+        CacheTuple cache(args.tree_name(), outputFile.get(), false);
+        auto summaryTuple = ntuple::CreateSummaryTuple("summary", originalFile.get(), true, ntuple::TreeState::Full);
+        summaryTuple->GetEntry(0);
+        SummaryInfo summaryInfo(summaryTuple->data(), Parse<Channel>(args.tree_name()), args.trigger_cfg());
+        for(const auto& event : *originalTuple) {
+            FillCacheTuple(cache, event,summaryInfo);
+            cache.Fill();
+        }
+
+        cache.Write();
+
+    }
+
+private:
+
+    void FillCacheTuple(CacheTuple& cacheTuple, const ntuple::Event& event,const SummaryInfo& summaryInfo) const
+    {
+        static const std::map<Channel, std::vector<std::string>> triggerPaths = {
+            { Channel::ETau, { "HLT_Ele32_WPTight_Gsf_v", "HLT_Ele35_WPTight_Gsf_v", "HLT_Ele24_eta2p1_WPTight_Gsf_LooseChargedIsoPFTau30_eta2p1_CrossL1_v" } },
+            { Channel::MuTau, { "HLT_IsoMu24_v", "HLT_IsoMu27_v", "HLT_IsoMu20_eta2p1_LooseChargedIsoPFTau27_eta2p1_CrossL1_v" } },
+            { Channel::TauTau, { "HLT_DoubleTightChargedIsoPFTau35_Trk1_TightID_eta2p1_Reg_v",
+                "HLT_DoubleMediumChargedIsoPFTau40_Trk1_TightID_eta2p1_Reg_v",
+                "HLT_DoubleTightChargedIsoPFTau40_Trk1_eta2p1_Reg_v"} },
+            { Channel::MuMu, { "HLT_IsoMu24_v", "HLT_IsoMu27_v" } },
+        };
+        const Channel channel = Parse<Channel>(args.tree_name());
+
+        if(event.extraelec_veto || event.extramuon_veto) return;
+
+        //JetOrdering jet_ordering = run_period == Period::Run2017 ? JetOrdering::DeepCSV : JetOrdering::CSV;
+        JetOrdering jet_ordering = JetOrdering::DeepFlavour;
+        std::set<size_t> Htt_indexes;
+        std::set<std::pair<size_t,size_t>> HH_indexes;
+        for(unsigned n = 0; n < signalObjectSelectors.size(); ++n){
+            for(unsigned h = 0; h < unc_sources.size(); ++h){
+                for(int l = -1; l < 2; ++l){
```

Comment:
what are h and l? Don't see any dependency on them below. Also note that `unc_scale = 0` for non central is not defined.